### PR TITLE
Floor OP

### PR DESCRIPTION
**Pull request by dc-dc-dc**
Adds floor op as per the suggestion in discord discussion, avoiding cast to int

Added tests for edge cases when values don't have a decimal

Tested on:
- CPU
- GPU
- METAL
- CUDA
- TORCH

Closes (#1038,  #998)

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/assembly_ptx.py           2               2
tinygrad/codegen/assembly_rdna.py          1               1
tinygrad/codegen/cstyle.py                 1               1
tinygrad/codegen/llvmir.py                 1               0
tinygrad/mlops.py                          4               1
tinygrad/ops.py                            1               1
tinygrad/runtime/ops_cpu.py                1               1
tinygrad/runtime/ops_torch.py              1               1
tinygrad/tensor.py                         4               4
------------------------------------------------------------
total                                     16              12
------------------------------------------------------------
lines added in the tinygrad folder: 4
```

File: tinygrad/codegen/assembly_rdna.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def specialize(self, asm) -> Tuple[str, str]:
     dtype_to_rdnatype = {dtypes.float32: "f32", dtypes.int64: "i64", dtypes.int32: "i32", dtypes.uint64: "u64", dtypes.bool: "i32"}
     alu = {BinaryOps.ADD: "add", BinaryOps.SUB: "sub", BinaryOps.MUL: "mul", FusedOps.MULACC: "fma",
            BinaryOps.MAX: "max", UnaryOps.RECIP: "rcp",
-           UnaryOps.NOOP: "mov", UnaryOps.SIN: "sin", UnaryOps.LOG2: "log", UnaryOps.EXP2: "exp",
+           UnaryOps.NOOP: "mov", UnaryOps.SIN: "sin", UnaryOps.LOG2: "log", UnaryOps.EXP2: "exp", UnaryOps.FLOOR: "trunc",
```
Comment (geohot) [inline tinygrad/codegen/assembly_rdna.py:48]: I don't think this is right, there's three "floor-like" operators. ceil, floor, and trunk. I think floor is floor.

**Comment by geohot** [comment]
So I'm mixed on this. Where is floor used in neural networks, and if it is being used, should it be?

I'm not sure where floor is used that a cast can't be. I understand the issue with large numbers, but when do you have those sorts of numbers and want to do a floor?

**Comment by dc-dc-dc** [comment]
It's mainly use in pre/post processing

**Comment by geohot** [comment]
Put it to a vote on Twitter :) I see arguments on both sides.

Pro:
* the int casting is a hack
* this fixes large numbers

Con:
* there's no derivative of floor
* all backends have to maintain this op now

https://twitter.com/__tinygrad__/status/1677359878929809408
